File: python/luhn/luhn.py
import copy

class Luhn:
    def __init__(self, card_num):
        try:
            self.numbers = [ int(num) for num in list( card_num.replace(' ', '') ) ]
        except:
            # it contains an invalid character and therefore doesn't matter
            self.numbers = None


    def valid(self):

        if( self.numbers == None or len(self.numbers) <= 1):
            return False

        numbers = copy.copy(self.numbers)
        numbers.reverse()
        
        for index, number in enumerate(numbers):
            if(index % 2 == 0):
                continue

            else:
                numbers[index] = number * 2 if number * 2 < 9 else (number * 2) - 9

        return sum(numbers) % 10 == 0

File: javascript/luhn/luhn.js

export const valid = (inputString) => {

  // remove whitespace
  inputString = inputString.split(' ').join('').split('');

  // invalid if too short
  if(inputString.length <= 1){
    return false;
  }

  // if there's an invalid character, it'll catch
  try{
    let numbers = inputString.map(elem => parseInt(elem));
    let reversedNumbers = numbers.slice().reverse();

    let sum = reversedNumbers.reduce((total, current, index) => {
      if(index % 2 == 1){
        return total + (current * 2 < 9 ? current * 2 : current * 2 - 9);
      }
      else{
        return total + current;
      }

    }, 0);



    return sum % 10 == 0;

  }
  catch(error){
    return false;
  }
};

